route-rs-packets/src/arp.rs

```diff
@@ -0,0 +1,253 @@
+use crate::{EthernetFrame, MacAddr};
+use std::convert::TryInto;
+use std::net::IpAddr;
+
+pub enum ArpOp {
+    Request = 1,
+    Reply = 2,
+}
+
+pub enum ArpHardwareType {
+    Ethernet = 1,
+}
+
+// TODO: should move to types.rs or ethernet.rs
+pub const IPV4_ETHER_TYPE: u16 = 0x0800;
+pub const IPV6_ETHER_TYPE: u16 = 0x86DD;
+pub const ARP_ETHER_TYPE: u16 = 0x0806;
+
+const HARDWARE_TYPE_RANGE: (usize, usize) = (0, 2);
+const PROTOCOL_TYPE_RANGE: (usize, usize) = (2, 4);
+const HARDWARE_ADDR_LEN_RANGE: (usize, usize) = (4, 5);
+const PROTOCOL_ADDR_LEN_RANGE: (usize, usize) = (5, 6);
+const OPCODE_RANGE: (usize, usize) = (6, 8);
+
+///
+/// EthernetFrame wrapper with getters/setters for the packet structure described in RFC 826
+/// https://tools.ietf.org/html/rfc826
+///
+// NOTE: Could be implemented in various ways, such as a specialized version of EthernetFrame that's
+// known to be an ARP frame. It could be implemented in a similar way that packets are promoted/demoted
+// with `TryFrom`.
+#[derive(Clone)]
+pub struct ArpFrame {
+    frame: EthernetFrame,
```

Comment:
Struct should have its own PacketData, and offset pointers.

Reply:
I actually disagree a bit here. I like that we already have an abstraction for dealing with header offsets and underlying data of a packet. `ArpFrame` is only supposed to encapsulate the details of the _payload_ of an ARP frame, but by unwrapping `EthernetFrame` we take on the additional complexity of knowing about the underlying buffer and necessary offsets. I'd rather hide that complexity inside of `EthernetFrame`.

Reply:
I guess another way to think about it is that `ArpFrame` isn't a primitive packet construct. An `ArpFrame` _is_ an `EthernetFrame` with additional constraints and ways to view its payload, so it seems like it should be composed of an `EthernetFrame` and use it as a client.

Reply:
Hmm, if you want to continue to wrap the EthernetFrame, that is fine, but it breaks from the way the rest of the Packets are laid out.  Also, tbf, I don't see many spots where you are you are actually using the EthernetFrame outside of reaching through it to the PacketData, or getting the payload/set_payload().  I'm not sure you are actually minimizing complexity by not stripping the PacketData and offsets into your own structure.  You also can't support the inside-out construction (encap functions) that other higher level Packets do, where you make a bare TCPHeader first, then encap it with IPv4, so on and so forth.  But if you think that the wrapped frame is the right way to go I trust your judgement since you have a better handle on how this is going to be used in the field than I do.